src/sync/record_handler.py

```diff
@@ -0,0 +1,78 @@
+"""
+Abstraction for New Content Handling
+provides a framework for writing new content to disk and posting to AWS
+"""
+from abc import ABC, abstractmethod
+
+from src.logger import set_log
+from src.models.block_range import BlockRange
+from src.sync.common import aws_login_and_upload
+from src.sync.config import SyncConfig
+
+log = set_log(__name__)
+
+
+class RecordHandler(ABC):  # pylint: disable=too-few-public-methods
+
+    """
+    This class is responsible for consuming new dune records and missing values from previous runs
+    it attempts to fetch content for them and filters them into "found" and "not found" as necessary
+    """
+
+    def __init__(
+        self,
+        block_range: BlockRange,
+        config: SyncConfig,
+    ):
+        self.config = config
+        self.block_range = block_range
+
+        self.name = config.table_name
+        self.file_path = config.volume_path / self.name
+        self.content_filename = f"cow_{block_range.block_to}.json"
+
+    @abstractmethod
+    def _num_records(self) -> int:
+        """Returns number of records to handle"""
+
+    @abstractmethod
+    def _write_found_content(self) -> None:
+        """Writes content to disk"""
+
+    @abstractmethod
+    def _write_sync_data(self) -> None:
+        """Records last synced content file"""
+
+    def write_and_upload_content(self, dry_run: bool) -> None:
+        """
+        - Writes self.order_rewards to persistent volume,
+        - attempts to upload to AWS and
+        - records last sync block on volume.
+        When dryrun flag is enabled, does not upload to IPFS.
+        """
+        self._write_found_content()
+        if self._num_records() > 0:
+            log.info(
+                f"attempting to post {self._num_records()} new {self.name} "
+                f"records for block range {self.block_range}"
+            )
+            if dry_run:
+                log.info(
+                    "DRY-RUN-ENABLED: New records written to volume, but not posted to AWS."
+                )
+            else:
+                aws_login_and_upload(
+                    config=self.config,
+                    path=self.file_path,
+                    filename=self.content_filename,
+                )
+                log.info(
+                    f"{self.name} sync for block range {self.block_range} complete: "
+                    f"synced {self._num_records()} records"
+                )
+        else:
+            log.info(
+                f"No new {self.name} for block range {self.block_range}: no sync necessary"
+            )
+
+        self._write_sync_data()
```

Comment:
Calling `_write_sync_data()` here looks strange. Is the intention here to let the base class call the deriving class's implementation of `_write_sync_data()` (because the base class's implementation is a no-op)?
Usually it only works the other way around (the derived class calls the implementation of the base class) and I would assume this is also how `abc` implements it.

Reply:
Maybe I just don't understand the abstraction. I would assume it calls the method. I'm pretty sure that's what it does since the abstraction forces the implementation to be there.

Reply:
Most of the time the point of abstract classes is to have some (optional) shared code together with some sort of interface. Then multiple classes can implement the same interface (by overriding the abstract methods) while being able to share the (optional) shared code of the abstract base class.
This however does not mean that the base class is able to call the code from classes that got derived from itself. Usually classes can only access their own code and the code from its parent but not from its children so to speak.

Reply:
From the `abc` docs:
```
Note
Unlike Java abstract methods, these abstract methods may have an implementation. This implementation can be called via the super() mechanism from the class that overrides it. This could be useful as an end-point for a super-call in a framework that uses cooperative multiple-inheritance.
```
And docs for [`super()`](https://docs.python.org/3/library/functions.html#super).

Reply:
So this is all fine and dandy, but I don't know what other pattern I am supposed to use instead. This code is doing what I expect -- do you happen to know what is the correct abstraction to be using for this purpose?

Reply:
What you could do is:
```
class RecordHandler {
def __init__(
        self,
        block_range: BlockRange,
        config: SyncConfig,
    ):
        self.config = config
        self.block_range = block_range

        self.name = config.table_name
        self.file_path = config.volume_path / self.name
        self.content_filename = f"cow_{block_range.block_to}.json"
        
    @abstractmethod
    def num_records(self) -> int:
        """Returns number of records to handle"""

    @abstractmethod
    def write_found_content(self) -> None:
        """Writes content to disk"""

    @abstractmethod
    def write_sync_data(self) -> None:
        """Records last synced content file"""
        
    def upload_to_aws(self, dry_run:bool) -> None:
               if dry_run:
                log.info(
                    "DRY-RUN-ENABLED: New records written to volume, but not posted to AWS."
                )
            else:
                aws_login_and_upload(
                    config=self.config,
                    path=self.file_path,
                    filename=self.content_filename,
                )
                log.info(
                    f"{self.name} sync for block range {self.block_range} complete: "
                    f"synced {self._num_records()} records"
                )
}
```

Then `AppDataHandler` and `OrderookDataHandler` could implement the abstract functions of `RecordHandler`.

And then you could have a `UploadHandler` which takes a `RecordHandler` which is now an interface which is implemented by `AppDataHandler` and `OrderookDataHandler` so you should be able to pass in either of those types into `UploadHandler`.
And finally `UploadHandler` manages the upload like this:
```
class UploadHandler
def __init__(handler: RecordHandler):
    self.handler = handler

def write_and_upload_content(self, dry_run: bool) -> None:
        """
        - Writes self.order_rewards to persistent volume,
        - attempts to upload to AWS and
        - records last sync block on volume.
        When dryrun flag is enabled, does not upload to IPFS.
        """
        super.write_found_content()
        if handler.num_records() > 0:
            log.info(
                f"attempting to post {handler.num_records()} new {self.name} "
                f"records for block range {self.block_range}"
            )
            handler.upload_to_aws()
        else:
            log.info(
                f"No new {self.name} for block range {self.block_range}: no sync necessary"
            )

        handler.write_sync_data()
```

This allows `AppDataHandler` and `OrderookDataHandler` to share some common code while also implementing some custom code and `UploadHandler` can just delegate to the concrete implementations.

Reply:
You could also make `RecordHandler` completely abstract and move `upload_to_aws` into `UploadHandler`. This is probably more correct.

Reply:
Thanks! I will do this!

Reply:
The only issue I see with this approach is the fact that I would have to make all the private methods public. They are meant to be private so that end users can not call them in the incorrect order. I am not so sure I want to do this...

Reply:
Okie - Thanks for this! Its much better now, although I think there is still room for some further generalizations I will save for later.

Reply:
Cool. Now it looks like it would call the expected code. 👍